src/defs.h

```diff
@@ -482,6 +483,7 @@ struct sockaddr_store_t {
         sa_family_t ss_family;
         sockaddr_in addr4;
         sockaddr_in6 addr6;
+        sockaddr_un addr_un;
```

Comment:
`sockaddr_un` is longer than 100 bytes so this structure size is 10x to previous revision. This may result in performance degradations in socket address handling code (e.g. statistics).

Reply:
@agalanin-at-nvidia should we pass this struct in another way? 
@rosenbaumalex  any ideas how to proceed?

Reply:
You need to carefully check all the places when address (sockaddr_store_t or IPAddress) is copied or used as a key in containers. At least you need to convert IPAddress to something memory-efficient than does not copy addr_un for IP addressed.

Reply:
@agalanin-at-nvidia  I guess it's important only in data path?  
if so, I can't find places where it's copy. 
AFAIK in both msg_sendto / receive_pending_data functions there is no copy of sockaddr_store_t.

Reply:
Some places I found:
* https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/server.h#L285
* https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/sockperf.cpp#L2556
* https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/client.h#L116
It's ironic that somewhere we have a comment "... to improve performance"

Reply:
`m_ServerList` map also copies `IPAddress` as a key. Probably we need to have three different maps for each socket address family (INET, INET6, UNIX) or variant type for map key that stores only 4 bytes for INET addresses etc.
https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/defs.h#L598
https://github.com/Mellanox/sockperf/blob/f80980feaf5334532d21df621c5791202f19dfab/src/client.h#L58

Reply:
I see that the union is 
```
class IPAddress {
    union {
        in_addr m_addr4;
        in6_addr m_addr6;
    };
```

Should I add 
`const char *m_addr_un;`  
or 
`sockaddr_un *m_addr_un`

Reply:
if it's marked as `const char` it can be added as 

```

  case AF_UNIX:
+            if (len >= sizeof(sockaddr_un)) {
+                m_addr_un = reinterpret_cast<const sockaddr_un *>(sa)->sun_path;
+            } else {
+                m_family = AF_UNSPEC;
+            }
+            break;

```

Reply:
Yes, idea with `char *` looks good. But not forget to strdup original `sun_path` on creation and free it in destructor.

Reply:
I thought more and realized that it is not so simple.

You must have full-length sockaddr for `recvfrom` because syscall known nothing about our optimizations. You also need to pass full sockaddr_un to `sendto` call. So I think that we must replace `sendto_addr = m_fds_ifd->server_addr;` with pointer assignment and move `char *` into `IPAddress` class.

Reply:
How can we optimize this?

Reply:
At first, check that we can use pointer instead of copy. If it is not possible, copy only needed bytes depending on the address family.

Reply:
Please take a loot at `copy_relevant_sockaddr_params` function in `common.h`